integration/aggregate_compat_test.go

```diff
@@ -0,0 +1,369 @@
+		},
+		"StageTypeMismatch": {
+			command: bson.D{
+				{"aggregate", "collection-name"},
+				{"pipeline", bson.A{1}},
+			},
+			resultType: emptyResult,
+		},
+	}
+
+	testAggregateCommandCompat(t, testCases)
+}
+
+func TestAggregateCompatStages(t *testing.T) {
+	testCases := map[string]aggregateStageCompatTestCase{
+		"MatchAndCount": {
+			pipeline: bson.A{
+				bson.D{{"$match", bson.D{{"v", 42}}}},
+				bson.D{{"$count", "v"}},
+				bson.D{{"$sort", bson.D{{"_id", 1}}}},
+			},
+		},
+	}
+
+	testAggregateStageCompat(t, testCases)
+}
+
+func TestAggregateCompatCount(t *testing.T) {
+	testCases := map[string]aggregateStageCompatTestCase{
+		"Value": {
+			pipeline: bson.A{bson.D{{"$count", "v"}}},
+		},
+		"NonExistent": {
+			pipeline: bson.A{bson.D{{"$count", "nonexistent"}}},
+		},
+		"Location15948": {
+			pipeline:   bson.A{bson.D{{"$count", "_id"}}},
+			resultType: emptyResult,
+		},
+		"Location40156": {
+			pipeline:   bson.A{bson.D{{"$count", 1}}},
+			resultType: emptyResult,
+		},
+		"Location40157": {
+			pipeline:   bson.A{bson.D{{"$count", ""}}},
+			resultType: emptyResult,
+		},
+		"Location40160": {
+			pipeline:   bson.A{bson.D{{"$count", "v.foo"}}},
+			resultType: emptyResult,
+		},
+		"Location40158": {
+			pipeline:   bson.A{bson.D{{"$count", "$foo"}}},
+			resultType: emptyResult,
+		},
+	}
+
+	testAggregateStageCompat(t, testCases)
+}
+
+func TestAggregateCompatMatch(t *testing.T) {
+	testCases := map[string]aggregateStageCompatTestCase{
+		"ID": {
+			pipeline: bson.A{bson.D{{"$match", bson.D{{"_id", "string"}}}}},
+		},
+		"Int": {
+			pipeline: bson.A{
+				bson.D{{"$match", bson.D{{"v", 42}}}},
+			},
+		},
+		"String": {
+			pipeline: bson.A{
+				bson.D{{"$match", bson.D{{"v", "foo"}}}},
+			},
+		},
+		"Document": {
+			pipeline: bson.A{bson.D{{"$match", bson.D{{"v", bson.D{{"foo", int32(42)}}}}}}},
+		},
+		"Array": {
+			pipeline: bson.A{bson.D{{"$match", bson.D{{"v", bson.A{int32(42)}}}}}},
+		},
+		"Regex": {
+			pipeline: bson.A{bson.D{{"$match", bson.D{{"v", bson.D{{"$eq", primitive.Regex{Pattern: "foo", Options: "i"}}}}}}}},
+		},
+		"Empty": {
+			pipeline: bson.A{
+				bson.D{{"$match", bson.D{}}},
+			},
+		},
+		"DotNotationDocument": {
+			pipeline: bson.A{
+				bson.D{{"$match", bson.D{{"v.foo", int32(42)}}}},
+			},
+		},
+		"DotNotationArray": {
+			pipeline: bson.A{
+				bson.D{{"$match", bson.D{{"v.0", int32(42)}}}},
+			},
+		},
+		"Location15959": {
+			pipeline:   bson.A{bson.D{{"$match", 1}}},
+			resultType: emptyResult,
+		},
+	}
+
+	testAggregateStageCompat(t, testCases)
+}
+
+func TestAggregateCompatSort(t *testing.T) {
+	testCases := map[string]aggregateStageCompatTestCase{
+		"AscendingID": {
+			pipeline: bson.A{bson.D{{"$sort", bson.D{{"_id", 1}}}}},
+		},
+		"DescendingID": {
+			pipeline: bson.A{bson.D{{"$sort", bson.D{{"_id", -1}}}}},
+		},
+		"AscendingValue": {
+			pipeline: bson.A{bson.D{{"$sort", bson.D{
+				{"v", 1},
+				{"_id", 1}, // sort by _id when v is the same.
```

Comment:
We could remove this, as it's the default test value, but I assume it's better to show it inside the test case as we test for `$sort` here particularly. :+1:

Reply:
So this one isn't default value, but `AscendingID` test it the default one. Let's keep it there because we would want to have a test case for it explicitly!